### ISSUE TITLE
Diff update of BRouter segments

### ISSUE THREAD

**Issue by Phyks** [abrensch/brouter#186]
Hi,

Currently, BRouter segments are generated weekly from the full planet dump at http://brouter.de/brouter/segments4/. One can also generate their own from a full planet dump.

However, the detailed way of generating them, using https://github.com/abrensch/brouter/blob/master/misc/scripts/mapcreation/process_pbf_planet.sh requires downloading the full planet dump for every generation.

Would it be feasible (or what would be needed) in order to be able to generate "diff" segments files, from the diffs (https://wiki.openstreetmap.org/wiki/Planet.osm/diffs)? OsmAnd for instance is now offering on the go update of the downloaded maps without a full download. This is motivating mappers to add road works and temporary restrictions, which BRouter could benefit from as well.

This would also have the nice advantage of ensuring the maps displayed and the route computed are somehow in sync.

Thanks

**Comment by abrensch** [comment]
you are interchanging 3 things:

1) generating rd5 from diffs: actually, I'm doing that. The weekly job does not use the weekly planet, but uses osmupdate ( https://wiki.openstreetmap.org/wiki/Osmupdate ) to update a local planet.pbf

2) feeding the users with rd5-delta's: well, possible, but quite some effort

3) generating rd5-deltas from osm-diff's is in some sense just the combination of the 2, but if you imply that this can be a fast process significantly below the 12 hours latency that the current process has, then the answer is no

**Comment by Phyks** [comment]
Thanks for the feedback on this. You are right there are multiple sides to this issue:

1. Generating faster the rd5 files (lower latency with real time data).
2. Providing smaller files to download to the BRouter app / clients.

Concerning 1, currently, rd5 files are generated from a full planet dump (either freshly downloaded or updated with osmupdate). I was wondering about generating rd5-delta's directly from diff files and whether this could have a lower latency that the current process. Based on your feedback, generating more than once a day the rd5s for the full planet is not really achievable (with reasonable computing power).

So, based on your feedback, 1 is just a matter of computing power and there is not much we can easily do to enhance this.

Concerning 2, assuming the rd5 files were somewhat continuously built (let's say once a day, I've no idea whether this could be reasonable/feasible with the current BRouter hosting infrastructure), this would mean a client would have to download the full file every day (about 50-100MB per rd5 segment).

However, the actual "diff" of the rd5 is probably much smaller than this. I don't know much about the rd5 format (is there a doc/readme somewhere?) and whether this format could easily be diff-compatible. The idea behind this being that if this could be achieved easily, BRouter app could probably be enabled to auto-download segments in the background and ensure the segments files are always up to date, even on a mobile connection.

**Comment by abrensch** [comment]
>> So, based on your feedback, 1 is just a matter of computing power and there
>> is not much we can easily do to enhance this.

The preprocessing chain was build with a hard memory constraint, had to work within 2,6 GB on Linux 32... With more memory you could streamline it to merge some steps or parallelize some others. But would it change anything if latency drops from 12h to 4h ? 

>> However, the actual "diff" of the rd5 is probably much smaller than this.
>> I don't know much about the rd5 format (is there a doc/readme somewhere?)

You don't need to understand rd5-format to do that:

There's an intermediate format that is now used only by the preprocessor: Serialisation of a node is first done into this intermediate format. If that is done for a complete tile (1/32 degree  * 1/32 degree), then this "MicroCache"  is "compressed" into rd5 format.

As an integrity check, the pre-processor then decodes this data back into a "MicroCache" and then compares if both versions are the same. If you look at that comparsion (  MicroCache._compareWith ) then you get an idea how Delta-RD5s could be created: 

- create Micro-Caches that contain only the data for updated/inserted/deleted nodes (where deleted would be represented as "empty" nodes with 0 data bytes). 

- build RD5 files the same way and in the same format as already done wthin "WayLinker", just using these "Delta-MicroCache-es". The codec just needs a small extension to be able to handle these "empty nodes", and we need to put reference version numbers somewhere in the header...

**Comment by abrensch** [comment]
I implemented a proof-of-concept for delta-rd5's ( https://github.com/abrensch/brouter/commit/a69fb1c99a8b164181eac1b4cf8cda09dcd13ac3 )

It's not yet creating delta-files, just doing the diff and the encoding, so that I can get the statistics.

In my test (weekly diff of E10_N50.rd5) I get the following stats:

   nodesTotal=5826549 nodesDiff=53525 bytesDiff=867652

So about 1% in node-count and in size. But this is including a minor lookup-change (from your PR), and that lookup change affects some 10k+ nodes in that region.

**Comment by Phyks** [comment]
> But would it change anything if latency drops from 12h to 4h ?

Indeed, this is not a major issue. Being able to build once per day is way more than enough. Not sure if this would be doable with the current infrastructure (latency is 12h, so it should be, I'm just wondering whether the server might have more to handle than just segments generation).

Building from diff files directly instead of handling the full planet file through `osmupdate` would only spare ressources, but this is probably the most interesting and necessary thing to work on at the moment indeed.

> nodesTotal=5826549 nodesDiff=53525 bytesDiff=867652

If I understand correctly, this means the matching `E10_N50.rd5` diff file would be about 800kB large (about 1% of the segments file as you say). This is really promising, as one usually needs only a few segments files and this would mean it is doable to have always up to date segments, downloading only a few MB per day!

**Comment by abrensch** [comment]
>>    But would it change anything if latency drops from 12h to 4h ?

> Indeed, this is not a major issue. Being able to build once per
> day is way more than enough. Not sure if this would be doable
> with the current infrastructure (latency is 12h, so it should be,
> I'm just wondering whether the server might have more to
> handle than just segments generation).

I tuned the preprocesser chain to 4h and switched to daily. The preprocessor is hosted by a server from the "historic.place" project, so I am just a guess, so eating up too much resources would be unfriendly.

I also extended the traffic simulation to include france, but at the same ztime made it a little bit sloppy by using the previous simulation result. So if you move a street, it will loose the "estimated_traffic_class" pseudo tag for one day.

No progress on the delta download of rd5 files. No problem either, just a complex task that's not on the top of the list.

**Comment by abrensch** [comment]
I completed the diffing infrastructure.

There's now a "diff" subdirectory of the segments4-folder that contains diff-files, e.g.:

  http://brouter.de/brouter/segments4/diff/E0_N40/

The name of the diff files contains the md5-hash of the base-file.

Creating the current rd5 from an old one and the diff file is done by e.g.:

java -cp brouter.jar  btools.mapaccess.Rd5DiffManager old/E0_N40.rd5 f78c81f18e706b51c3844799292619d0.rd5diff E0_N40.rd5 

This is done automatically by the download-manager of the brouter-app (in the upcoming release)

Normally a daily diff is about 0,5% the size of a data file.

However, there's still a problem with the tiles that contain the traffiic-Simulation.

There seems to be too much noise from the simulation, so there the daily diff is about 5%. Bigger probleme here is that about 50% of the micro-tiles a "dirty", leading to up to 10 Minutes cpu-time to reconstruct a big file.

So I leave this issue here open to track the problem of traffic simuation noise.
